### PR TITLE
static jax structures & dispersive adjoint

### PR DESCRIPTION
**Pull request by tylerflex**
Support for dispersive materials in adjoint (with shape / boundary optimization).
Adds two new adjoint structure classes: 
* `JaxStructureStaticMedium` combines a jax geometry with a tidy3d medium.
* `JaxStructureStaticGeometry` combines a tidy3d geometry with a jax medium.

The former can be used to pair a differentiable geometry (`JaxPolySlab`) with any medium, eg a dispersive medium or something from the material library (like a metal).

This PR mainly generalizes the `JaxGeometry` internals to an abstract class and then splits into the three `JaxStructure`, `JaxStructureStaticMedium`, and `JaxStructureStaticGeometry`to handle the field definitions and to do different adjoint handling depending on their `_differentiable_fields` class attribute.

Among another static geometry structure, the tests have a `JaxStructureStaticMedium` containing a `JaxBox` and silver from our material library. So I think we can begin using this for eg metallic antenna design.

### CODE REVIEW

File: tests/test_plugins/test_adjoint.py [hunk 1]
```diff
@@ -253,6 +257,14 @@ def make_sim(
 
     jax_geo_group = JaxGeometryGroup(geometries=[jax_polyslab1, jax_polyslab1])
     jax_struct_group = JaxStructure(geometry=jax_geo_group, medium=jax_med1)
+
+    jax_struct_static_med = JaxStructureStaticMedium(
+        geometry=jax_box1, medium=td.Medium()  # material_library["Ag"]["Rakic1998BB"]
```
Comment (momchil-flex) [inline tests/test_plugins/test_adjoint.py:262]: Did you forget to uncomment the silver?

**Review by momchil-flex — APPROVED** [review]
Awesome. Somewhat related, and also related to our discussion yesterday, we should (or do we already?) validate that `JaxSimulations` do not have nonlinear media in either their static structures or their JaxStructureStaticMedium ones.

**Comment by momchil-flex** [comment]
I guess warning could be fine too yeah.

**Comment by e-g-melo** [comment]
Hi @tylerflex!

I tried the `JaxStructureStaticMedium` and `JaxStructureStaticGeometry` within the adjoint examples 5 and 3, respectively, and both finished with an error (16:54:02 -03 ERROR: Error running task!).

In the `JaxStructureStaticMedium` class, the error happened when inserting both a non-dispersive or a dispersive medium (['Si3N4']['Philipp1973Sellmeier']): (task id)[https://tidy3d.simulation.cloud/workbench?taskId=fdve-db1d154d-fd94-4477-944d-ea989bb748dd]

In the `JaxStructureStaticGeometry` test, I've replaced the `JaxBox` by a `Box` object, and the same error has happened: (task id)[https://tidy3d.simulation.cloud/workbench?taskId=fdve-4758b5bf-188b-4e67-85ab-f5a5d48cd833].

**Comment by tylerflex** [comment]
This is probably because we'd need to build a solver to use this front end commit. Can you try with a run_local instead of run?

**Comment by tylerflex** [comment]
Note this PR adds a new field to the Jax info datastructure to remember what kind of input structures were used. So we need to make sure this is handled in the backend too. It might not even work with run local. @momchil-flex if it doesn't maybe could you build a solver using this commit so i can test?

**Comment by e-g-melo** [comment]
Ok.. got it!

**Comment by momchil-flex** [comment]
I've built a release and associated it to frontend version `2.5.0`, modify `version.py` to use.

**Comment by tylerflex** [comment]
Thanks!

**Review by e-g-melo — APPROVED** [review]
Hi, @tylerflex and @momchil-flex!

It is working now! The gradients were calculated with no errors in both examples.

Thanks for this feature Tyler! Taking into account dispersive materials in the simulation adds another level of interest to the adjoint.

**Comment by e-g-melo** [comment]
Sure, I will run the complete optimizations and update them in docs.

**Comment by tylerflex** [comment]
Thanks Emerson. @momchil-flex ill do nonlinear warnings in a separate PR

**Pull request merged**